libbeat/processors/add_cloud_metadata/provider_aws_ec2.go

```diff
@@ -18,125 +18,135 @@
+	return ec2.NewFromConfig(cfg)
+}
 
-	tokenReq, err := http.NewRequest("PUT", fmt.Sprintf("http://%s%s", metadataHost, ec2InstanceIMDSv2TokenURI), nil)
-	if err != nil {
-		logger.Warnf("error when make token request for getting IMDSv2 token: %s. No token in the metadata request will be used.", err)
-		return ""
-	}
+// AWS EC2 Metadata Service
+var ec2MetadataFetcher = provider{
+	Name: "aws-ec2",
 
-	tokenReq.Header.Add(ec2InstanceIMDSv2TokenTTLHeader, ec2InstanceIMDSv2TokenTTLValue)
-	rsp, err := client.Do(tokenReq)
-	if rsp == nil {
-		logger.Warnf("read token request for getting IMDSv2 token returns empty: %s. No token in the metadata request will be used.", err)
-		return ""
-	}
+	Local: true,
 
-	defer func(body io.ReadCloser) {
-		if body != nil {
-			body.Close()
+	Create: func(_ string, config *conf.C) (metadataFetcher, error) {
+		ec2Schema := func(m map[string]interface{}) mapstr.M {
+			m["service"] = mapstr.M{
+				"name": "EC2",
+			}
+			return mapstr.M{"cloud": m}
 		}
-	}(rsp.Body)
 
+		fetcher, err := newGenericMetadataFetcher(config, "aws", ec2Schema, fetchRawProviderMetadata)
+		return fetcher, err
+	},
+}
+
+// fetchRaw queries raw metadata from a hosting provider's metadata service.
+func fetchRawProviderMetadata(
+	ctx context.Context,
+	client http.Client,
+	result *result,
+) {
+	logger := logp.NewLogger("add_cloud_metadata")
+
+	// LoadDefaultConfig loads the EC2 role credentials
+	awsConfig, err := awscfg.LoadDefaultConfig(context.TODO(), awscfg.WithHTTPClient(&client))
 	if err != nil {
-		logger.Warnf("error when read token request for getting IMDSv2 token: %s. No token in the metadata request will be used.", err)
-		return ""
+		logger.Debugf("error loading AWS default configuration: %s.", err)
+		result.err = errors.Wrapf(err, "failed loading AWS default configuration")
+		return
 	}
+	awsClient := NewIMDSClient(awsConfig)
 
-	if rsp.StatusCode != http.StatusOK {
-		logger.Warnf("error when check request status for getting IMDSv2 token: http request status %d. No token in the metadata request will be used.", rsp.StatusCode)
-		return ""
+	instanceIdentity, err := awsClient.GetInstanceIdentityDocument(context.TODO(), &imds.GetInstanceIdentityDocumentInput{})
+	if err != nil {
+		logger.Debugf("error fetching EC2 Identity Document: %s.", err)
+		result.err = errors.Wrapf(err, "failed fetching EC2 Identity Document.")
+		return
 	}
 
-	all, err := ioutil.ReadAll(rsp.Body)
+	// AWS Region must be set to be able to get EC2 Tags
+	awsRegion := instanceIdentity.InstanceIdentityDocument.Region
+	awsConfig.Region = awsRegion
+
+	clusterName, err := fetchEC2ClusterNameTag(awsConfig, instanceIdentity.InstanceIdentityDocument.InstanceID)
 	if err != nil {
-		logger.Warnf("error when reading token request for getting IMDSv2 token: %s. No token in the metadata request will be used.", err)
-		return ""
+		logger.Debugf("error fetching cluster name metadata: %s.", err)
```

Comment:
This can be Warnf level an not debug?

Reply:
@gizas thank you for pointing it out! From my understanding it should be fine to use it this way, http fetcher uses this definition as well - https://github.com/elastic/beats/blob/98b5cb5a60108be18ba94d4af4dc3af199e64c1c/libbeat/processors/add_cloud_metadata/http_fetcher.go#L94

Reply:
My comment has to do more with the level error vs debug vs warn. I would say not to use debug as it would hide any messages unless you raise the log level

Reply:
this code is outdated - was replaced to `Warn` in https://github.com/elastic/beats/pull/35182/commits/0241f2270f774f17f9a345d393bfba28219550d8